File: pset6/morecomf/credit.py
# import sys
import sys
import cs50

# main starts here
def main():

    cc = int(input("Enter the credit card number: "))

    n , odd, even = 0, 0, 0

    # checksum
    while True:
        last = (cc % 10)
        cc = int(cc / 10)
        if n % 2 != 0:
            add = last * 2
            even += int((add / 10)) + (add % 10)
        else:
            odd += last

        n += 1

        if (cc <= 9):
            break

    # checksum for first number of the cc
    if n % 2 != 0:
        add = cc * 2
        even += int((add / 10)) + (add % 10)
    else:
        odd += cc


    check = (cc * 10) + last

    # check if the card is valid
    if (odd + even) % 10 == 0:
        if cc == 4:
            print("Visa")
        elif check == 34 or check == 37:
            print("AMEX")
        elif check == 51 or check == 52 or check == 53 or check == 54 or check == 55:
            print("MasterCard")
        else:
            print("invalid card")
    else:
        print("invalid card")


if __name__ == "__main__":
    main()

File: pset1/morecomf/credit.c
#include<stdio.h>
#include<cs50.h>

// returns remainder
int splitx(long long ccnum)
{
    int junk = ccnum % 10;
    return junk;
}

// main starts here
int main(void)
{
    printf("Number:\t");
    long long ccnum=get_long_long();
    int ultimate = 0, n = 0, decimal = 0, sum = 0, sum1 = 0, sum2 = 0;

    // iterate though all the numbers
    do
    {
        decimal = ultimate;
        ultimate = ccnum;
        int junk = splitx(ccnum);
        if(n % 2 == 0)
            sum1 = sum1 + junk;
        else
        {
            int junks = junk * 2;
            sum2 = sum2 + (junks % 10) + (junks / 10);
        }
        ccnum /= 10;
        n++;
    }
    while(ccnum > 1);

    // check checksum
    sum = sum1 + sum2;
    if(sum % 10 == 0)
    {
        // determine which company distributed the card
        if (decimal == 34 || decimal == 37)
            printf("AMEX\n");
        if (ultimate == 4)
            printf("VISA\n");
        if (decimal ==51 || decimal == 52 || decimal == 53 || decimal == 54 || decimal == 55)
            printf("MASTERCARD\n");
    }
    else
        printf("INVALID\n");
    return 0;
}